src/config/resolver.rs

```diff
@@ -0,0 +1,96 @@
+use std::collections::VecDeque;
+
+use crate::config::is_default;
+use crate::config::{Config, Source};
+use anyhow::anyhow;
+use serde::{Deserialize, Serialize};
+use tokio::fs::File;
+use tokio::io::AsyncReadExt;
+use url::Url;
+
+#[derive(Default, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+enum LinkType {
+  #[default]
+  Config,
+  GraphQL,
+  Protobuf,
+  Data,
+}
+
+#[derive(Default, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct Link {
+  #[serde(default, skip_serializing_if = "is_default")]
+  type_of: LinkType, // Type of the link
+  #[serde(default, skip_serializing_if = "is_default")]
+  src: String, // Source URL for linked files
+  #[serde(default, skip_serializing_if = "is_default")]
+  id: Option<String>, // Id is used to refer at different places in the config
+  #[serde(default, skip_serializing_if = "is_default")]
+  content: Option<String>, // Stores raw content
+}
+
+impl Link {
+  pub async fn resolve_recurse(self) -> anyhow::Result<Vec<Link>> {
+    let mut result: Vec<Link> = Vec::new();
+    if self.type_of == LinkType::Protobuf || self.type_of == LinkType::Data {
+      let link: Link = Self::get_raw_content(&self).await?;
+      result.push(link);
+      return Ok(result);
+    }
+    let mut link_queue: VecDeque<Link> = VecDeque::new();
+    link_queue.push_back(self);
+    while let Some(mut curr) = link_queue.pop_front() {
+      let (txt, source) = if let Ok(url) = Url::parse(&curr.src) {
+        let resp = reqwest::get(url).await?;
+        let path = curr.src.clone();
+        if !resp.status().is_success() {
+          return Err(anyhow!("Read over URL failed with status code: {}", resp.status()));
+        }
+        let source = if let Some(v) = resp.headers().get("content-type") {
+          if let Ok(s) = Source::detect_content_type(v.to_str()?) {
+            s
+          } else {
+            Source::detect(path.trim_end_matches('/'))?
+          }
+        } else {
+          Source::detect(path.trim_end_matches('/'))?
+        };
```

Comment:
likely we can make the condition response an `Option` enum so we can make `let source = source.or(Source::detect(path.trim_end_matches('/'))?)` instead of returning `Source::detect(path.trim_end_matches('/'))?` on `else` blocks.

Reply:
another option is to put this logic into a private function.

so you can make something like
```suggestion
        let source = get_source(resp).or(Source::detect(path.trim_end_matches('/'))?;
```

Reply:
and the private function would be closer to this:
```
  fn get_source(resp: /* the type */) -> anyhow::Result<Source> {
    let content_type = resp.headers().get("content-type")?;

    Source::detect_content_type(content_type)
  }
```

note this is not a ready-to-go code, and it'll likely require a few tweaks